### ISSUE TITLE
Generate Clang-toolchain based compile_commands.json on Windows host

### ISSUE THREAD

**Issue by ifarbod** [xmake-io/xmake#4559]
### Xmake Version

2.8.6

### Operating System Version and Architecture

Windows 11 23H2

### Describe Bug

I have a basic project definition.

XMake works fine for the platform I compile for (MSVC), however, I use VSCode with the official Clangd extension, and would like to use XMake to generate `compile_commands.json` - it does that just fine, however, Clangd gives better analysis results (especially for warnings) if the compile_commands are in clang/clang++ format, XMake however, generates the JSON file with MSVC (cl.exe) in mind. I don't want this behavior and there's no way to turn it around currently (not that I know of).

### Expected Behavior

`set_toolchains("clang")` globally without any condition works just fine, but I only want it to work only after executing `xmake project -k compile_commands`

Currently it gives me:

```json
[
{
  "directory": "c:\\$Files\\Projects\\xmake-bug",
  "arguments": ["C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.38.33130\\bin\\HostX86\\x86\\cl.exe", "/c", "/nologo", "/Zi", "/FS", "/Fdbuild\\windows\\x86\\debug\\compile.xmake-bug.pdb", "/Od", "/std:c++latest", "/Isrc", "/EHsc", "/Fobuild\\.objs\\xmake-bug\\windows\\x86\\debug\\src\\Main.cpp.obj", "src\\Main.cpp", "-imsvc", "C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.22621.0\\cppwinrt", "-imsvc", "C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.22621.0\\km", "-imsvc", "C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.22621.0\\shared", "-imsvc", "C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.22621.0\\ucrt", "-imsvc", "C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.22621.0\\um", "-imsvc", "C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.22621.0\\winrt", "-imsvc", "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.38.33130\\include"],
  "file": "src\\Main.cpp"
}]
```

But I want:

```json
[
{
  "directory": "C:\\$Files\\Projects\\xmake-bug",
  "arguments": ["clang.exe", "-c", "-Qunused-arguments", "-m32", "-g", "-O0", "-std=c++23", "-Isrc", "-fexceptions", "-fcxx-exceptions", "-o", "build\\.objs\\xmake-bug\\windows\\x86\\debug\\src\\Main.cpp.obj", "src\\Main.cpp"],
  "file": "src\\Main.cpp"
}]

```

### Project Configuration

set_xmakever("2.8.6")

set_languages("cxxlatest")
set_allowedplats("windows")
set_arch("x86")

-- this is the issue
if is_config("compile_commands") then
    set_toolchains("clang")
end

add_rules("mode.debug", "mode.releasedbg", "mode.release")
add_rules("plugin.vsxmake.autoupdate")

target("xmake-bug")
    set_kind("binary")
    add_files("src/**.cpp")
    add_includedirs("src/")

### Additional Information and Error Logs

There are no errors.

**Comment by SirLynix** [comment]
You can generate a compile_commands.json with Clangd in mind using `--lsp=clangd`.

`xmake project -k compile_commands --lsp=clangd`

**Comment by ifarbod** [comment]
@SirLynix That still generates cl.exe lines

**Comment by ifarbod** [comment]
I have also tried this, but doesn't have any effect:

```lua
rule("mode.clangd")
    on_config(function (target)
        print("Running compile_commands")
        target:set("toolchains", "clang")
    end)
rule_end()

add_rules("mode.debug", "mode.releasedbg", "mode.release", "mode.final", "mode.clangd")
```

`$ xmake project -k compile_commands -m clangd`

**Comment by waruqi** [comment]
It works for me.

```lua
add_rules("mode.debug", "mode.release")

set_toolchains("clang")

target("test5")
    set_kind("binary")
    add_files("src/*.cpp")
```

```console
>xmake project -k compile_commands
create ok!

[
{
  "directory": "D:\\projects\\personal\\test5",
  "arguments": ["clang.exe", "-c", "-Qunused-arguments", "-m64", "-fvisibility=hidden", "-fvisibility-inlines-hidden", "-O3", "-fexceptions", "-fcxx-exceptions", "-DNDEBUG", "-o", "build\\.objs\\test5\\windows\\x64\\release\\src\\main.cpp.obj", "src\\main.cpp"],
  "file": "src\\main.cpp"
}]

```

**Comment by waruqi** [comment]
> if is_config("compile_commands") then
    set_toolchains("clang")
end

you can remove `is_config("compile_commands")`

**Comment by SirLynix** [comment]
I think they want to compile with MSVC but use clangd as a linter in VS Code.

**Comment by ifarbod** [comment]
@waruqi I think a flag to pass to the project subcommand would be an ideal solution:

```shell
$ xmake project -k compile_commands --cc=clang
```

Or making the `is_config()` check true or implementing`get_config()` return something that isn't `nil`.

**Comment by waruqi** [comment]
> @waruqi I think a flag to pass to the project subcommand would be an ideal solution:
> 
> ```shell
> $ xmake project -k compile_commands --cc=clang
> ```
> 
> Or making the `is_config()` check true or implementing`get_config()` to return something that isn't `nil`.

please define `option("compile_commands")` and use `has_config` instead of `is_config`, then run

```console
xmake f --compile_commands=y
xmake project -k compile_commands
```

```lua
option("compile_commands", {default = false, description = "xxxx"})
if has_config("compile_commands") then
    set_toolchains("clang")
end
```

**Comment by waruqi** [comment]
you can see https://xmake.io/#/plugin/more_plugins?id=configure-xmakelua-to-generate-automatically

**Comment by ifarbod** [comment]
@waruqi No, the point is that I'm already using the so-called "[Automatic trigger generation](https://xmake.io/#/plugin/more_plugins?id=automatic-trigger-generation)" but since it doesn't have the flag, it'll override it back to MSVC, I need XMake to have a way to differentiate between normal project generation, building, and `compile_commands.json` updating.

To make it more clear, I use Clangd, not Microsoft's cpptools. XMake assumes anyone that uses Windows as their build host _**and**_ target is using cpptools by default, which makes sense as that's what most people use and clangd is less popular.

**Comment by waruqi** [comment]
I still don't know what you're trying to do. 

1. If you want to generate compile_comands.json using the clang compiler, please use `set_toolchains("clang")`, it works for me. see https://github.com/xmake-io/xmake/issues/4559#issuecomment-1874765312
2. If you want to generate a clangd-style compile_commands.json, set lsp to clangd
3. If you want to automatically update compile_commands.json, please use `plugin.compile_commands.autoupdate` rule
4. If you want to use clangd and configure compile_commands.json, please see https://xmake.io/#/plugin/more_plugins?id=clangd

```lua
set_toolchains("clang")
add_rules("plugin.compile_commands.autoupdate", {outputdir = ".vscode", lsp = "clangd"})
target("test")
     set_kind("binary")
     add_files("src/*.c")
```

**Comment by SirLynix** [comment]
To be more clear as I think I understood what @ifarbod want.

They want to use msvc to compile on Windows **and** to use clangd as a linter (in the compile_commands.json), and they want their compile_commands to automatically update with the rule, without touching the project config as it wouldn't work otherwise.

So they want their toolchain to be the default one, but to generate (and automatically update) compile_commands with clangd toolchain.

**Comment by waruqi** [comment]
> To be more clear as I think I understood what @ifarbod want.
> 
> They want to use msvc to compile on Windows **and** to use clangd as a linter (in the compile_commands.json), and they want their compile_commands to automatically update with the rule, without touching the project config as it wouldn't work otherwise.
> 
> So they want their toolchain to be the default one, but to generate (and automatically update) compile_commands with clangd toolchain.

xmake-vscode can set `xmake.compileCommandsBackend` to clangd, and it has been default value.

https://github.com/xmake-io/xmake-vscode/blob/a5bdf61a9ff0053ddb2bc7380e330495e2dd77ba/package.json#L680

so, we need not do any thing. just use default config + vscode-xmake

```lua
target("test")
    add_files("src/*.cpp")
```

**Comment by SirLynix** [comment]
Yes, but passing `--lsp=clangd` (which is the same thing as setting `XMAKE_GENERATOR_COMPDB_LSP`) keeps generating cl.exe command-line:
```
[
{
  "directory": "c:\\$Files\\Projects\\xmake-bug",
  "arguments": ["C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.38.33130\\bin\\HostX86\\x86\\cl.exe", "/c", "/nologo", "/Zi", "/FS", "/Fdbuild\\windows\\x86\\debug\\compile.xmake-bug.pdb", "/Od", "/std:c++latest", "/Isrc", "/EHsc", "/Fobuild\\.objs\\xmake-bug\\windows\\x86\\debug\\src\\Main.cpp.obj", "src\\Main.cpp", "-imsvc", "C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.22621.0\\cppwinrt", "-imsvc", "C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.22621.0\\km", "-imsvc", "C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.22621.0\\shared", "-imsvc", "C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.22621.0\\ucrt", "-imsvc", "C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.22621.0\\um", "-imsvc", "C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.22621.0\\winrt", "-imsvc", "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.38.33130\\include"],
  "file": "src\\Main.cpp"
}]
```

**Comment by waruqi** [comment]
> They want to use msvc to compile on Windows **and** to use clangd as a linter (in the compile_commands.json)

> They want to use msvc to compile on Windows and to use clangd as a linter (in the compile_commands.json)


You just said compiling with msvc, so even though you use clangd, you're still using the msvc compiler.

Unless you want to use clangd and also use clang in compile_commands.json.

`--lsp=clangd` only modifies the compile_commands generation style to make it better supported by clangd. But it does not modify the compiler inside compile_commands.

**Comment by SirLynix** [comment]
> You just said compiling with msvc, so even though you use clangd, you're still using the msvc compiler.
>
> Unless you want to use clangd and also use clang in compile_commands.json.

Yes that's the issue of @ifarbod, he wants to compile using MSVC (when running xmake) but have a compile_commands generated with Clang (for Visual Studio Code linter).

**Comment by ifarbod** [comment]
> Unless you want to use clangd and also use clang in compile_commands.json.

Yes, that's what I want. Please provide a built-in method _or_ guide me to the responsible code so I can make a PR.

Before discovering XMake, this has been me and my team's standard workflow for writing C++, granted, i was using Premake5 with a bunch of Python scripts.

I never knew `compile_commands` with cl.exe entries was possible, after switching to XMake, I was disappointed to see that. It's not useful to anyone that doesn't use vscode-cpptools, because they use it as pseudo-makefiles to build the project (?) and have Microsoft's own IntelliSense engine index the code using it. (I am aware, they also have clangd support)

For me and anyone else that prefers `vscode-clangd`, cl.exe entries are not that useful.
This extension is more like a linter instead of a full on IDE experience (as @SirLynix mentioned). To make the most use of it, you'd provide it with clang's [objectively better] warning definitions and have it complain at your code.

**Comment by waruqi** [comment]
try

```lua
target("test")
    set_kind("binary")
    add_files("src/*.cpp")
    on_load(function (target)
        if os.getenv("XMAKE_IN_PROJECT_GENERATOR") then
            target:set("toolchains", "clang")
        end
    end)
```

**Comment by ifarbod** [comment]
@waruqi It keeps reverting back to cl.exe

**Comment by waruqi** [comment]
you can try it. But it may have some side effects (caching causes). We can't support it better at the moment.

```lua
target("test")
    set_kind("binary")
    add_files("src/*.cpp")
    after_load(function (target)
        if os.getenv("XMAKE_IN_PROJECT_GENERATOR") then
            target:set("toolchains", "clang")
        end
        for _, toolchain in ipairs(target:toolchains()) do
            toolchain:check()
        end
    end)
```

**Comment by ifarbod** [comment]
@waruqi Tested this, it's okay-ish for the time being, hopefully, there's a better solution in the future.

**Comment by ifarbod** [comment]
2 years later, can we revisit this?
